packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.ts

```diff
@@ -109,17 +109,23 @@ export class CryptoWatchPriceFeed extends PriceFeedInterface {
     // If the time is before the first piece of data in the set, return null because
     // the price is before the lookback window.
     if (time < firstPricePeriod.openTime - this.historicalTimestampBuffer) {
-      throw new Error(`${this.uuid}: time ${time} is before firstPricePeriod.openTime`);
+      throw new Error(`${this.uuid}: time ${time} is before firstPricePeriod.openTime minus historicalTimestampBuffer`);
     }
 
     // historicalPricePeriods are ordered from oldest to newest.
-    // This finds the first pricePeriod whose closeTime is after the provided time.
-    const match = this.historicalPricePeriods.find((pricePeriod) => {
-      return (
-        time < pricePeriod.closeTime + this.historicalTimestampBuffer &&
-        time >= pricePeriod.openTime - this.historicalTimestampBuffer
-      );
+    // This finds the first pricePeriod whose closeTime is after the provided time. First try to find the match without
+    // resorting to using the historical timestamp buffer, then if there is no match found try using the buffer.
+    let match = this.historicalPricePeriods.find((pricePeriod) => {
+      return time < pricePeriod.closeTime && time >= pricePeriod.openTime;
     });
+    if (match === undefined && this.historicalTimestampBuffer > 0) {
+      match = this.historicalPricePeriods.find((pricePeriod) => {
```

Comment:
Nice catch.

One oddity remaining is that if this buffer were large, let's say an hour, it would grab the fall back to the _first_ valid timestamp, meaning if the data point didn't exist it would fall back to the one one hour ago rather than the price period just before the missing one.

Does that make sense? And can you think of any easy way to fix this?

Reply:
In my specific use case fork here https://github.com/Reinis-FRP/protocol/blob/75eb066e5180409964991d02a642ecd65c5e2853/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js#L132 I have separately searched for nearest timestamp backwards or nearest timestamp forward. Though in my use case I afterwards use only prior close price, but it could be extended to use also forward open price if needed.

Reply:
@Reinis-FRP do you think its a mistake in your fork that you first try to `match` using the buffer in [this line](https://github.com/Reinis-FRP/protocol/blob/75eb066e5180409964991d02a642ecd65c5e2853/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js#L124) and then if that fails then you try to match `before`? It seems like no matter what if the `time` is within a `pricePeriodTimestamp - historicalTimestampBuffer` that you'll match the earliest price (but not the nearest)?

Reply:
Say the price periods have the following timestamps:

```
[
     {openTime: 0, closeTime: 1, price: 1}
     {openTime: 1, closeTime: 2, price: 2}
     {openTime: 2, closeTime: 3, price: null}
     {openTime: 3, closeTime: 4, price: 4}
]
```

And say we're trying to find `getHistoricalTime(time=2)`. Let `historicalTimestampBuffer=1`. In your fork it seems that you'll match the first time period where `price=1` because of this code:

```
    const match = this.historicalPricePeriods.find((pricePeriod) => {
      return (
        time < pricePeriod.closeTime + this.historicalTimestampBuffer &&
        time >= pricePeriod.openTime - this.historicalTimestampBuffer
      );
    });
```

But the correct price period to match will be 2, right?

Reply:
I did not change that part - its the same [here](https://github.com/UMAprotocol/protocol/blob/65600aec5f53ae17792b4191a8dd03a03deceba7/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.ts#L119). It always seemed weird to me why the first timestamp should be taken, but did not fully understand the intended use case of `historicalTimestampBuffer`. So I did not use it and instead created `maxBeforeTimestampBuffer` and if time was not within any open-close period then returned first matching price backwards if it was within `maxBeforeTimestampBuffer`.